proposals/infra/INF-0003-packoffset.md

```diff
@@ -0,0 +1,141 @@
+<!-- {% raw %} -->
+
+# Strict Availability Diagnostics
+
+* Proposal: [INF-0003](INF-0003-packoffset.md)
+* Author(s): [Xiang Li](https://github.com/python3kgae)
+* Sponsor: 
+* Status: **Under Consideration**
+* Planned Version: 
+
+## Introduction
+
+packoffset attribute is used to change the layout of a cbuffer.
+
+```
+packoffset( c[Subcomponent][.component] )
+```
+
+It will overwrite the layout of a cbuffer with the packoffset attributes on
+the constants.
+
+Here are several examples of manually packing shader constants.
+
+```
+cbuffer MyBuffer
+{
+    float4 Element1 : packoffset(c0);
+    float1 Element2 : packoffset(c12);
+    float1 Element3 : packoffset(c1.y);
+}
+```
+without the packoffset, Element2 will be in c1.x.
+
+
+### Limitations
+
+There're some limitations when apply packoffset on a constant.
+
+1. packoffset is only allowed in a constant buffer.
+```
+float c : packoffset(c0); // invalid for apply packoffset on a global constant.
+```
+
+2. cannot mix packoffset elements with nonpackoffset elements in a cbuffer.
+```
+cbuffer A {
+  float a : packoffset(c0);
+  float b;  // invalid for mix packoffset elements and nonpackoffset elements.
+}
+```
+3. Pack subcomponents of vectors and scalars whose size is large enough to prevent crossing register boundaries.
+```
+cbuffer A {
+  float2 a : packoffset(c0.w); // invalid for a.x in c0.w and a.y in c1.x 
+                               // which crossed register boundary.
+}
+```
+4. struct, matrix, array cannot have component.
+```
+struct S {
+  float a;
+};
+
+cbuffer A{
+  S s : packoffset(c2.y); // invalid for offset to component y
+  int a[2] : packoffset(c3.z); // invalid for offset to component z
+  float2x2 m : packoffset(c6.z); // invalid for offset to component z
+  S s2 : packoffset(c12); // valid
+  int a2[2] : packoffset(c13); // valid
+  float2x2 m2 : packoffset(c16); // valid
+}
+```
+
+## Motivation
+
+To support all HLSL features available in compute profile.
+
+## Proposed solution
+
+### AST
+
+A new attribute HLSLPackOffsetAttr will be created with:
+```
+def HLSLPackOffset: HLSLAnnotationAttr {
+  let Spellings = [HLSLAnnotation<"packoffset">];
+  let LangOpts = [HLSL]
+  let Args = [IntArgument<"RegNum">, IntArgument<"Component", /*optional*/ 1>]
+}
+```
+
+The AST for MyBuffer example will be
+```
+HLSLBufferDecl cbuffer MyBuffer
+  VarDecl Element1 float4
+    HLSLPackOffsetAttr   0 0
+  VarDecl Element2 float
+    HLSLPackOffsetAttr  12 0
+  VarDecl Element3 float
+    HLSLPackOffsetAttr   1 1
+  
+```
+
+### llvm IR
+
+To apply the packoffset to a cbuffer, the layout of the struct constructed for 
+the cbuffer requires reordering of its fields, and padding should be introduced
+if any gaps exist between the fields.
+```
+cbuffer CB {
+  float a : packoffset(c2);
+  float b : packoffset(c4);
+  float2 c : packoffset(c2.z);
+}
+```
+will get layout struct like
+```
+struct CBLayout {
+  float4 padding0[2]; // c0/c1
```

Comment:
With padding like this I wonder how something like this would work out:

```
cbuffer A {
  float 4 B : packoffset(c1);
}

float4 C : register(c0);
```

Reply:
Or similarly:
````
cbuffer A {
    float4 B : packoffset(c1);
}

cbuffer C {
  float4 D : register(c0);
}
````

Reply:
They're in different cbuffer.
The C is in global CB where B is in cbuffer A.